shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -16,23 +16,55 @@
 
 package com.nvidia.spark.rapids.shims.spark310
 
+import java.nio.{ByteBuffer, ByteOrder}
+
 import scala.collection.JavaConverters._
 import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
 
 import ai.rapids.cudf._
 import com.nvidia.spark.rapids._
 import com.nvidia.spark.rapids.GpuColumnVector.GpuColumnarBatchBuilder
 import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import org.apache.avro.util.ByteBufferInputStream
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.HadoopReadOptions
+import org.apache.parquet.hadoop.{ParquetFileReader, ParquetInputFormat}
+import org.apache.parquet.io.{DelegatingSeekableInputStream, InputFile, SeekableInputStream}
 
 import org.apache.spark.TaskContext
 import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen.GenerateUnsafeProjection
 import org.apache.spark.sql.columnar.{CachedBatch, CachedBatchSerializer}
+import org.apache.spark.sql.execution.datasources.parquet.{ParquetReadSupport, ParquetToSparkSchemaConverter, ParquetWriteSupport}
+import org.apache.spark.sql.execution.datasources.parquet.rapids.ParquetRecordMaterializer
 import org.apache.spark.sql.internal.SQLConf
-import org.apache.spark.sql.types.{StructField, StructType}
+import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
+import org.apache.spark.sql.types._
 import org.apache.spark.sql.vectorized.ColumnarBatch
 import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.SerializableConfiguration
+
+class ByteArrayInputFile(buff: Array[Byte]) extends InputFile {
+
+  override def getLength: Long = buff.length
+
+  override def newStream(): SeekableInputStream = {
+    // ParquetFileReader reads in little endian! :shrugs
+    val byteBuffer = ByteBuffer.wrap(buff).order(ByteOrder.LITTLE_ENDIAN)
```

Comment:
Yes, this is necessary because ParquetFileReader is [reading](https://github.com/apache/parquet-mr/blob/a89df8f9932b6ef6633d06069e50c9b7970bebd1/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L520) the footer in LittleEndian while the default byte order in Java is Big-Endian. 
In a Big-Endian environment, the order of the ByteBuffer created on line 57 will still be Little-endian so I don't see how this will affect. Please let me know what I am missing.

Reply:
In the case of the footer, the Parquet reader is reading a byte stream a byte at a time and explicitly interpreting the bytes as they arrive as little-endian via `readIntLittleEndian`.  There is no endian-ness when a byte stream is read in this manner.  Endian only matters when multiple bytes are loaded in a single transaction, not byte-at-a-time.

Think of how this works in the filesystem.  It's a stream of bytes we write to the filesystem, and we don't have to tell the Parquet file reader whether the file is little endian or big endian, so why would we need to do something different here?  The stream of bytes should be able to be passed to the Parquet reader and it's up to that reader to decide how to interpret the byte stream, just like it interprets the byte stream from a file.

Reply:
This is my fault. I walked back in my local history and found a version where I was reading from a different stream than where I was updating the position. Fate has it that the number being read in was very close to a byte shifted version, then I read the `readIntLittleEndian` method and that gave me a clue that may be it was expecting the byte order to be different from what I am sending it. Making the change to the byte order involved creating a ByteBuffer, it was during this change that I inadvertently fixed the original bug which made me think this actually fixed the problem but I still didn't quite understand and that reflected in my comment. Bravo to @jlowe for picking up on that subtlety *bow*